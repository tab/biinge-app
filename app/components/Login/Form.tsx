import React, { useEffect, useState } from "react"
import { View, TextInput, Keyboard } from "react-native"
import { FormikErrors, FormikProps, withFormik } from "formik"
import { useTranslation } from "react-i18next"

import i18n from "config/i18n"
import InputError from "components/ui/InputError"
import Button from "components/ui/Button"
import { EMAIL_REGEX, PASSWORD_LENGTH } from "helpers/validation"
import { LoginFormValues } from "types"
import { loginFormStyles } from "styles"
import colors from "styles/colors"

interface FormProps {
  initialValues: LoginFormValues
  isLoading: boolean
  onSubmit: (values: LoginFormValues) => void
}

type Props = FormikProps<LoginFormValues> & FormProps

const LoginForm = ({
  isLoading,
  errors,
  setFieldValue,
  handleSubmit,
}: Props) => {
  const { t } = useTranslation()

  const [visible, setVisible] = useState(false)

  useEffect(() => {
    const showSubscription = Keyboard.addListener("keyboardWillShow", () => {
      setVisible(true)
    })
    const hideSubscription = Keyboard.addListener("keyboardWillHide", () => {
      setVisible(false)
    })

    return () => {
      showSubscription.remove()
      hideSubscription.remove()
    }
  }, [])

  return (
    <View
      style={
        visible
          ? [loginFormStyles.content, loginFormStyles.contentPadding]
          : loginFormStyles.content
      }
    >
      <View style={loginFormStyles.group}>
        <TextInput
          testID="email-input"
          autoCapitalize="none"
          autoComplete="email"
          autoCorrect={false}
          autoFocus={true}
          allowFontScaling={false}
          textContentType="emailAddress"
          style={loginFormStyles.input}
          placeholder={t("login.form.email.placeholder.title")}
          placeholderTextColor={colors.gray}
          onChangeText={(newValue: string) => {
            setFieldValue("email", newValue)
          }}
        />
        <InputError>{errors.email}</InputError>
      </View>
      <View style={loginFormStyles.group}>
        <TextInput
          testID="password-input"
          autoCapitalize="none"
          autoComplete="password"
          autoCorrect={false}
          autoFocus={false}
          allowFontScaling={false}
          secureTextEntry
          textContentType="password"
          style={loginFormStyles.input}
          placeholder={t("login.form.password.placeholder.title")}
          placeholderTextColor={colors.gray}
          onChangeText={(newValue: string) => {
            setFieldValue("password", newValue)
          }}
        />
        <InputError>{errors.password}</InputError>
      </View>
      <View style={loginFormStyles.actions}>
        <Button
          testID="login-button"
          style={
            isLoading
              ? [loginFormStyles.submit, loginFormStyles.disabled]
              : [loginFormStyles.submit]
          }
          disabled={isLoading}
          onPress={handleSubmit}
        >
          {t("login.form.submit.title")}
        </Button>
      </View>
    </View>
  )
}

const LoginFormWithFormik = withFormik<FormProps, LoginFormValues>({
  enableReinitialize: true,
  mapPropsToValues: ({ initialValues }) => initialValues,
  handleSubmit: (values, { props: { onSubmit } }) => onSubmit(values),
  validate: (values: LoginFormValues) => {
    const errors: FormikErrors<LoginFormValues> = {}

    const { email, password } = values

    if (!email) {
      errors.email = i18n.t("validation.field.required", {
        field: i18n.t("login.form.email.placeholder.title"),
      })
    } else if (!EMAIL_REGEX.test(email)) {
      errors.email = i18n.t("validation.field.invalid", {
        field: i18n.t("login.form.email.placeholder.title"),
      })
    }

    if (!password) {
      errors.password = i18n.t("validation.field.required", {
        field: i18n.t("login.form.password.placeholder.title"),
      })
    } else if (password.length < PASSWORD_LENGTH) {
      errors.password = i18n.t("validation.password.toShort")
    }

    return errors
  },
})(LoginForm)

export default LoginFormWithFormik
