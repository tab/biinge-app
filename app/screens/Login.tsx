import React, { useEffect, useState } from "react"
import {
  View,
  Image,
  KeyboardAvoidingView,
  Pressable,
  Keyboard,
  Platform,
} from "react-native"
import FlashMessage, { showMessage } from "react-native-flash-message"
import LinearGradient from "react-native-linear-gradient"
import { AuthOperationName, useEmailPasswordAuth } from "@realm/react"

import i18n from "config/i18n"
import From from "components/Login/Form"
import { loginStyles, flashStyles, loginFormStyles } from "styles"
import { LoginFormValues } from "types"
import colors from "styles/colors"

const LoginScreen = () => {
  const { logIn, register, result } = useEmailPasswordAuth()

  const [values, setValues] = useState({ email: "", password: "" })

  useEffect(() => {
    if (result.error) {
      if (result.operation === AuthOperationName.LogIn) {
        register(values)

        setTimeout(() => {
          handleError(i18n.t("login.error.title"))
        }, 500)
      }

      if (result.operation === AuthOperationName.Register) {
        handleError(i18n.t("login.error.title"))
      }
    }

    // NOTE: login after registration
    if (result.success && result.operation === AuthOperationName.Register) {
      logIn(values)
    }
  }, [result, values])

  const handleError = (message: string) => {
    showMessage({
      backgroundColor: colors.darkCandyAppleRed,
      color: colors.white,
      message: message,
      type: "default",
    })
  }

  const handleClick = () => {
    Keyboard.dismiss()
  }

  const handleSubmit = (values: LoginFormValues) => {
    setValues(values)
    logIn(values)
  }

  return (
    <View style={loginStyles.root}>
      <FlashMessage
        position="top"
        textStyle={flashStyles.text}
        titleStyle={flashStyles.title}
      />
      <Image
        style={loginStyles.background}
        source={require("../assets/background.png")}
      />
      <LinearGradient
        style={loginStyles.background}
        colors={[
          "rgba(0, 0, 0, 0) 100%)",
          "rgba(0, 0, 0, 0.6) 50%",
          "rgba(0, 0, 0, 1) 0%",
        ]}
      />
      <KeyboardAvoidingView
        behavior={Platform.OS === "ios" ? "padding" : "height"}
        style={loginFormStyles.root}
      >
        <Pressable style={loginStyles.button} onPress={handleClick} />
        <From
          initialValues={{ ...values }}
          isLoading={result.pending}
          onSubmit={handleSubmit}
        />
      </KeyboardAvoidingView>
    </View>
  )
}

export default LoginScreen
