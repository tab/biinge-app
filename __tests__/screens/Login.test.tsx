import React from "react"
import "react-native"
import { render, fireEvent, act } from "@testing-library/react-native"

import { it } from "@jest/globals"

import Login from "screens/Login"

jest.mock("@realm/react", () => ({
  useEmailPasswordAuth: () => ({
    logIn: jest.fn(),
    register: jest.fn(),
    result: { pending: false, success: false, error: null, operation: null },
  }),
}))

describe("Login", () => {
  it("renders correctly", () => {
    const tree = render(<Login />)
    expect(tree).toMatchSnapshot()
  })

  it("handles backdrop click event", () => {
    const { getByTestId } = render(<Login />)
    fireEvent.press(getByTestId("login-backdrop"))
  })

  it("handles form submit", async () => {
    const { getByTestId } = render(<Login />)

    await act(async () => {
      fireEvent.changeText(getByTestId("email-input"), "test@example.com")
      fireEvent.changeText(getByTestId("password-input"), "password123")
      fireEvent.press(getByTestId("login-button"))
    })
  })
})
