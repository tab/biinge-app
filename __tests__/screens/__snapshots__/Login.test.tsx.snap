// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": [
          {
            "translateY": -48,
          },
        ],
        "zIndex": 99,
      }
    }
  />
  <Image
    source={
      {
        "testUri": "../../../app/assets/background.png",
      }
    }
    style={
      {
        "bottom": 0,
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "width": "100%",
      }
    }
  />
  <BVLinearGradient
    colors={
      [
        0,
        2566914048,
        4278190080,
      ]
    }
    endPoint={
      {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    startPoint={
      {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      {
        "bottom": 0,
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "width": "100%",
      }
    }
  />
  <View
    onLayout={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "transparent",
          "flex": 1,
          "width": "100%",
        }
      }
      testID="login-backdrop"
    />
    <View
      style={
        {
          "paddingBottom": 60,
          "paddingHorizontal": 24,
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "marginBottom": 15,
          }
        }
      >
        <TextInput
          allowFontScaling={false}
          autoCapitalize="none"
          autoComplete="email"
          autoCorrect={false}
          autoFocus={true}
          keyboardAppearance="light"
          onChangeText={[Function]}
          placeholder="Email"
          placeholderTextColor="#b5b5b5"
          style={
            {
              "backgroundColor": "#ffffff",
              "borderRadius": 4,
              "fontSize": 16,
              "padding": 12,
            }
          }
          testID="email-input"
          textContentType="emailAddress"
        />
      </View>
      <View
        style={
          {
            "marginBottom": 15,
          }
        }
      >
        <TextInput
          allowFontScaling={false}
          autoCapitalize="none"
          autoComplete="password"
          autoCorrect={false}
          autoFocus={false}
          keyboardAppearance="light"
          onChangeText={[Function]}
          placeholder="Password"
          placeholderTextColor="#b5b5b5"
          secureTextEntry={true}
          style={
            {
              "backgroundColor": "#ffffff",
              "borderRadius": 4,
              "fontSize": 16,
              "padding": 12,
            }
          }
          testID="password-input"
          textContentType="password"
        />
      </View>
      <View
        style={
          {
            "flexDirection": "column",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "backgroundColor": "#2b2835",
                "borderRadius": 50,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 15,
              },
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#181818",
                  "borderRadius": 4,
                  "justifyContent": "center",
                  "padding": 15,
                },
              ],
            ]
          }
          testID="login-button"
        >
          <Text
            style={
              [
                {
                  "color": "#ffffff",
                  "fontSize": 18,
                  "fontWeight": "bold",
                  "textAlign": "center",
                },
                undefined,
              ]
            }
          >
            Login or Register
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
